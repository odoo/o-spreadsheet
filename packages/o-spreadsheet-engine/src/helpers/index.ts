export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
