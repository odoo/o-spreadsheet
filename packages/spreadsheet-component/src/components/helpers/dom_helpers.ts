import { Rect } from "./../../types/rendering";

const macRegex = /Mac/i;

const MODIFIER_KEYS = ["Shift", "Control", "Alt", "Meta"];

/**
 * Return true if the event was triggered from
 * a child element.
 */
export function isChildEvent(parent: HTMLElement | null | undefined, ev: Event): boolean {
  if (!parent) return false;
  return !!ev.target && parent!.contains(ev.target as Node);
}

export function gridOverlayPosition() {
  const spreadsheetElement = document.querySelector(".o-grid-overlay");
  if (spreadsheetElement) {
    const { top, left } = spreadsheetElement?.getBoundingClientRect();
    return { top, left };
  }
  throw new Error("Can't find spreadsheet position");
}

export function getBoundingRectAsPOJO(el: Element): Rect {
  const rect = el.getBoundingClientRect();
  return {
    x: rect.x,
    y: rect.y,
    width: rect.width,
    height: rect.height,
  };
}

/**
 * Iterate over all the children of `el` in the dom tree starting at `el`, depth first.
 */
export function* iterateChildren(el: Node): Generator<Node> {
  yield el;
  if (el.hasChildNodes()) {
    for (let child of el.childNodes) {
      yield* iterateChildren(child);
    }
  }
}

export function getOpenedMenus(): HTMLElement[] {
  return Array.from(document.querySelectorAll<HTMLElement>(".o-spreadsheet .o-menu"));
}

const letterRegex = /^[a-zA-Z]$/;

/**
 * Transform a keyboard event into a shortcut string that represent this event. The letters keys will be uppercased.
 *
 * @argument ev - The keyboard event to transform
 * @argument mode - Use either ev.key of ev.code to get the string shortcut
 *
 * @example
 * event : { ctrlKey: true, key: "a" } => "Ctrl+A"
 * event : { shift: true, alt: true, key: "Home" } => "Alt+Shift+Home"
 */
export function keyboardEventToShortcutString(
  ev: KeyboardEvent,
  mode: "key" | "code" = "key"
): string {
  let keyDownString = "";
  if (!MODIFIER_KEYS.includes(ev.key)) {
    if (isCtrlKey(ev)) keyDownString += "Ctrl+";
    if (ev.altKey) keyDownString += "Alt+";
    if (ev.shiftKey) keyDownString += "Shift+";
  }
  const key = mode === "key" ? ev.key : ev.code;
  keyDownString += letterRegex.test(key) ? key.toUpperCase() : key;
  return keyDownString;
}

export function isMacOS(): boolean {
  return Boolean(macRegex.test(navigator.userAgent));
}

/**
 * @param {KeyboardEvent | MouseEvent} ev
 * @returns Returns true if the event was triggered with the "ctrl" modifier pressed.
 * On Mac, this is the "meta" or "command" key.
 */
export function isCtrlKey(ev: KeyboardEvent | MouseEvent): boolean {
  return isMacOS() ? ev.metaKey : ev.ctrlKey;
}
