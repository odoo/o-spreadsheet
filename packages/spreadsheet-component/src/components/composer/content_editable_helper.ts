import { deepEquals, toHex } from "../../helpers";
import { iterateChildren } from "../helpers/dom_helpers";
import { NEWLINE } from "./../../constants";
import { HtmlContent } from "./composer/composer";

export class ContentEditableHelper {
  // todo make el private and expose dedicated methods
  el: HTMLElement;
  constructor(el: HTMLElement) {
    this.el = el;
  }

  updateEl(el: HTMLElement) {
    this.el = el;
  }

  /**
   * select the text at position start to end, no matter the children
   */
  selectRange(start: number, end: number) {
    let selection = window.getSelection()!;
    const { start: currentStart, end: currentEnd } = this.getCurrentSelection();

    if (currentStart === start && currentEnd === end) {
      return;
    }
    const currentRange = selection.getRangeAt(0);
    let range: Range;
    if (this.el.contains(currentRange.startContainer)) {
      range = currentRange;
    } else {
      range = document.createRange();
      selection.removeAllRanges();
      selection.addRange(range);
    }
    if (start === end && start === 0) {
      range.setStart(this.el, 0);
      range.setEnd(this.el, 0);
    } else {
      const textLength = this.getText().length;
      if (start < 0 || end > textLength) {
        console.warn(
          `wrong selection asked start ${start}, end ${end}, text content length ${textLength}`
        );
        if (start < 0) start = 0;
        if (end > textLength) end = textLength;
        if (start > textLength) start = textLength;
      }
      let startNode = this.findChildAtCharacterIndex(start);
      let endNode = this.findChildAtCharacterIndex(end);
      range.setStart(startNode.node, startNode.offset);
      selection.extend(endNode.node, endNode.offset);
    }
  }

  /**
   * finds the dom element that contains the character at `offset`
   */
  private findChildAtCharacterIndex(offset: number): { node: Node; offset: number } {
    let it = iterateChildren(this.el);
    let current, previous;
    let usedCharacters = offset;
    let isFirstParagraph = true;
    do {
      current = it.next();
      if (!current.done && !current.value.hasChildNodes()) {
        if (current.value.textContent && current.value.textContent.length < usedCharacters) {
          usedCharacters -= current.value.textContent.length;
        } else if (
          current.value.textContent &&
          current.value.textContent.length >= usedCharacters
        ) {
          it.return(current.value);
        }
        previous = current.value;
      }
      // One new paragraph = one new line character, except for the first paragraph
      if (!current.done && current.value.nodeName === "P") {
        if (isFirstParagraph) {
          isFirstParagraph = false;
        } else {
          usedCharacters--;
        }
      }
    } while (!current.done && usedCharacters);

    if (current.value) {
      return { node: current.value, offset: usedCharacters };
    }
    return { node: previous, offset: usedCharacters };
  }

  /**
   * Sets (or Replaces all) the text inside the root element in the form of distinctive paragraphs and
   * span for each element provided in `contents`.
   *
   * The function will apply the diff between the current content and the new content to avoid the systematic
   * destruction of DOM elements which interferes with IME[1]
   *
   * Each line of text will be encapsulated in a paragraph element.
   * Each span will have its own fontcolor and specific class if provided in the HtmlContent object.
   *
   * [1] https://developer.mozilla.org/en-US/docs/Glossary/Input_method_editor
   */
  setText(contents: HtmlContent[][]) {
    if (contents.length === 0) {
      this.removeAll();
      return;
    }

    const childElements = Array.from(this.el.childNodes);
    const contentLength = contents.length;

    for (let i = 0; i < contentLength; i++) {
      const line = contents[i];
      const childElement = childElements[i];

      let newChild = false;
      let p: HTMLParagraphElement;
      if (childElement && childElement.nodeName === "P") {
        p = childElement as HTMLParagraphElement;
      } else {
        newChild = true;
        p = document.createElement("p");
      }

      const lineLength = line.length;
      const existingChildren = Array.from(p.childNodes);

      for (let j = 0; j < lineLength; j++) {
        const content = line[j];
        const child = existingChildren[j];
        // child nodes can be multiple types of nodes: Span, Text, Div, etc...
        // We can only modify a node in place if it has the same type as the content
        // that we would insert, which are spans.
        // Otherwise, it means that the node has been input by the user, through the keyboard or a copy/paste
        // @ts-ignore (somehow required because jest does not like child.tagName despite the prior check)
        const childIsSpan = child && "tagName" in child && child.tagName === "SPAN";
        if (childIsSpan && compareContentToSpanElement(content, child as HTMLSpanElement)) {
          continue;
        }
        // this is an empty line in the content
        if (!content.value && !content.class) {
          if (child) p.removeChild(child);
          continue;
        }
        const span = document.createElement("span");
        span.innerText = content.value;
        span.style.color = content.color || "";
        if (content.class) {
          span.classList.add(content.class);
        }
        if (child) {
          p.replaceChild(span, child);
        } else {
          p.appendChild(span);
        }
      }

      if (existingChildren.length > lineLength) {
        for (let i = lineLength; i < existingChildren.length; i++) {
          p.removeChild(existingChildren[i]);
        }
      }

      // Empty line
      if (!p.hasChildNodes()) {
        const span = document.createElement("span");
        span.appendChild(document.createElement("br"));
        p.appendChild(span);
      }

      // replace p if necessary
      if (newChild) {
        if (childElement) {
          this.el.replaceChild(p, childElement);
        } else {
          this.el.appendChild(p);
        }
      }
    }
    if (childElements.length > contentLength) {
      for (let i = contentLength; i < childElements.length; i++) {
        this.el.removeChild(childElements[i]);
      }
    }
  }

  scrollSelectionIntoView() {
    const focusedNode = document.getSelection()?.focusNode;
    if (!focusedNode || !this.el.contains(focusedNode)) return;
    const element = focusedNode instanceof HTMLElement ? focusedNode : focusedNode.parentElement;
    element?.scrollIntoView({ block: "nearest" });
  }

  /**
   * remove the current selection of the user
   * */
  removeSelection() {
    let selection = window.getSelection()!;
    selection.removeAllRanges();
  }

  private removeAll() {
    if (this.el) {
      while (this.el.firstChild) {
        this.el.removeChild(this.el.firstChild);
      }
    }
  }

  /**
   * finds the indexes of the current selection.
   * */
  getCurrentSelection() {
    let { startElement, endElement, startSelectionOffset, endSelectionOffset } =
      this.getStartAndEndSelection();
    let startSizeBefore = this.findSelectionIndex(startElement!, startSelectionOffset);
    let endSizeBefore = this.findSelectionIndex(endElement!, endSelectionOffset);

    return {
      start: startSizeBefore,
      end: endSizeBefore,
    };
  }

  /**
   * Computes the text 'index' inside this.el based on the currently selected node and its offset.
   * The selected node is either a Text node or an Element node.
   *
   * case 1 -Text node:
   * the offset is the number of characters from the start of the node. We have to add this offset to the
   * content length of all previous nodes.
   *
   * case 2 - Element node:
   * the offset is the number of child nodes before the selected node. We have to add the content length of
   * all the bnodes prior to the selected node as well as the content of the child node before the offset.
   *
   * See the MDN documentation for more details.
   * https://developer.mozilla.org/en-US/docs/Web/API/Range/startOffset
   * https://developer.mozilla.org/en-US/docs/Web/API/Range/endOffset
   *
   */
  private findSelectionIndex(nodeToFind: Node, nodeOffset: number): number {
    let usedCharacters = 0;

    let it = iterateChildren(this.el);
    let current = it.next();
    let isFirstParagraph = true;
    while (!current.done && current.value !== nodeToFind) {
      if (!current.value.hasChildNodes()) {
        if (current.value.textContent) {
          usedCharacters += current.value.textContent.length;
        }
      }
      // One new paragraph = one new line character, except for the first paragraph
      if (
        current.value.nodeName === "P" ||
        (current.value.nodeName === "DIV" && current.value !== this.el) // On paste, the HTML may contain <div> instead of <p>
      ) {
        if (isFirstParagraph) {
          isFirstParagraph = false;
        } else {
          usedCharacters++;
        }
      }
      current = it.next();
    }
    if (current.value !== nodeToFind) {
      /** This situation can happen if the code is called while the selection is not currently on the ContentEditableHelper.
       * In this case, we return 0 because we don't know the size of the text before the selection.
       *
       * A known occurence is triggered since the introduction of commit d4663158 (PR #2038).
       *
       * FIXME: find a way to test eventhough the selection API is not available in jsDOM.
       */
      return 0;
    } else {
      if (!current.value.hasChildNodes()) {
        usedCharacters += nodeOffset;
      } else {
        const children = [...current.value.childNodes].slice(0, nodeOffset);
        usedCharacters += children.reduce((acc: number, child: Node, index: number) => {
          if (child.textContent !== null) {
            // need to account for paragraph nodes that implicitely add a new line
            // except for the last paragraph
            let chars = child.textContent.length;
            if (child.nodeName === "P" && index !== children.length - 1) {
              chars++;
            }
            return acc + chars;
          } else {
            return acc;
          }
        }, 0);
      }
    }
    if (nodeToFind.nodeName === "P" && !isFirstParagraph && nodeToFind.textContent === "") {
      usedCharacters++;
    }
    return usedCharacters;
  }

  private getStartAndEndSelection() {
    const selection = document.getSelection()!;

    return {
      startElement: selection.anchorNode || this.el,
      startSelectionOffset: selection.anchorOffset,
      endElement: selection.focusNode || this.el,
      endSelectionOffset: selection.focusOffset,
    };
  }

  getText(): string {
    let text = "";

    let it = iterateChildren(this.el);
    let current = it.next();
    let isFirstParagraph = true;
    while (!current.done) {
      if (!current.value.hasChildNodes()) {
        text += current.value.textContent;
      }
      if (
        current.value.nodeName === "P" ||
        (current.value.nodeName === "DIV" && current.value !== this.el) // On paste, the HTML may contain <div> instead of <p>
      ) {
        if (isFirstParagraph) {
          isFirstParagraph = false;
        } else {
          text += NEWLINE;
        }
      }
      current = it.next();
    }
    return text;
  }
}

function compareContentToSpanElement(content: HtmlContent, node: HTMLElement): boolean {
  const contentColor = content.color ? toHex(content.color) : "";
  const nodeColor = node.style?.color ? toHex(node.style.color) : "";

  const sameColor = contentColor === nodeColor;
  const sameClass = deepEquals([content.class], [...node.classList]);
  const sameContent = node.innerText === content.value;
  return sameColor && sameClass && sameContent;
}
