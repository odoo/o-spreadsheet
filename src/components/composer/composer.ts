import * as owl from "@odoo/owl";
import { fontSizeMap } from "../../fonts";
import { composerTokenize, EnrichedToken, rangeReference } from "../../formulas/index";
import { colors, DEBUG, getComposerSheetName } from "../../helpers/index";
import { Rect, SpreadsheetEnv, Zone } from "../../types/index";
import { TextValueProvider } from "./autocomplete_dropdown";
import { ContentEditableHelper } from "./content_editable_helper";

const Component = owl.Component;
const { useRef, useState } = owl.hooks;
const { xml, css } = owl.tags;

export const FunctionColor = "#4a4e4d";
export const OperatorColor = "#3da4ab";
export const StringColor = "#f6cd61";
export const NumberColor = "#02c39a";
export const MatchingParenColor = "pink";

const tokenColor = {
  OPERATOR: OperatorColor,
  NUMBER: NumberColor,
  STRING: StringColor,
  BOOLEAN: NumberColor,
  FUNCTION: FunctionColor,
  DEBUGGER: OperatorColor,
  LEFT_PAREN: OperatorColor,
  RIGHT_PAREN: OperatorColor,
};

const TEMPLATE = xml/* xml */ `
<div class="o-composer-container" t-att-style="containerStyle">
    <div class="o-composer"
      t-att-style="composerStyle"
      t-ref="o_composer"
      tabindex="1"
      contenteditable="true"
      spellcheck="false"

      t-on-keydown="onKeydown"
      t-on-input="onInput"
      t-on-keyup="onKeyup"

      t-on-blur="saveSelection"
      t-on-click="onClick"
    />
    <TextValueProvider
        t-if="autoCompleteState.showProvider"
        t-ref="o_autocomplete_provider"
        search="autoCompleteState.search"
        provider="autoCompleteState.provider"
        t-on-completed="onCompleted"
    />
</div>
  `;
const CSS = css/* scss */ `
  .o-composer-container {
    box-sizing: border-box;
    position: absolute;
    padding: 0;
    margin: 0;
    border: 0;
    z-index: 5;
    .o-composer {
      caret-color: black;
      box-sizing: border-box;
      background-color: white;
      padding-left: 2px;
      padding-right: 2px;
      border: 1.6px solid #3266ca;
      white-space: nowrap;
      &:focus {
        outline: none;
      }
    }
  }
`;

export class Composer extends Component<any, SpreadsheetEnv> {
  static template = TEMPLATE;
  static style = CSS;
  static components = { TextValueProvider };

  private composerRef = useRef("o_composer");
  private autoCompleteRef = useRef("o_autocomplete_provider");

  getters = this.env.getters;
  dispatch = this.env.dispatch;

  zone: Zone;
  rect: Rect;

  selectionEnd: number = 0;
  selectionStart: number = 0;
  contentHelper: ContentEditableHelper;

  autoCompleteState = useState({
    showProvider: false,
    provider: "functions",
    search: "",
  });
  tokenAtCursor: EnrichedToken | void = undefined;

  // we can't allow input events to be triggered while we remove and add back the content of the composer in processContent
  shouldProcessInputEvents: boolean = false;
  // a composer edits a single cell. After that, it is done and should not
  // modify the model anymore.
  isDone: boolean = false;
  refSelectionStart: number = 0;
  tokens: EnrichedToken[] = [];

  keyMapping: { [key: string]: Function } = {
    Enter: this.processEnterKey,
    Escape: this.processEscapeKey,
    Tab: (ev: KeyboardEvent) => this.processTabKey(ev),
    F2: () => console.warn("Not implemented"),
    F4: () => console.warn("Not implemented"),
    ArrowUp: this.processArrowKeys,
    ArrowDown: this.processArrowKeys,
    ArrowLeft: this.processArrowKeys,
    ArrowRight: this.processArrowKeys,
  };

  constructor() {
    super(...arguments);
    this.contentHelper = new ContentEditableHelper(this.composerRef.el!);
    const [col, row] = this.getters.getPosition();
    this.zone = this.getters.expandZone({ left: col, right: col, top: row, bottom: row });
    this.rect = this.getters.getRect(this.zone, this.props.viewport);
  }

  mounted() {
    DEBUG.composer = this;

    const el = this.composerRef.el!;

    this.contentHelper.updateEl(el);
    const currentContent = this.getters.getCurrentContent();
    if (currentContent) {
      this.contentHelper.insertText(currentContent);
      this.contentHelper.selectRange(currentContent.length, currentContent.length);
    }
    this.processContent();

    el.style.width = (Math.max(el.scrollWidth + 10, this.rect[2] + 0.5) + "px") as string;
    el.style.height = (this.rect[3] + 0.5 + "px") as string;
  }

  willUnmount(): void {
    delete DEBUG.composer;
    this.trigger("composer-unmounted");
  }

  get containerStyle() {
    const style = this.getters.getCurrentStyle();
    const [x, y, , height] = this.rect;
    const weight = `font-weight:${style.bold ? "bold" : 500};`;
    const italic = style.italic ? `font-style: italic;` : ``;
    const strikethrough = style.strikethrough ? `text-decoration:line-through;` : ``;
    return `left: ${x - 1}px;
        top:${y}px;
        height:${height}px;
        font-size:${fontSizeMap[style.fontSize || 10]}px;
        ${weight}${italic}${strikethrough}`;
  }

  get composerStyle() {
    const style = this.getters.getCurrentStyle();
    const cell = this.getters.getActiveCell() || { type: "text" };
    const height = this.rect[3];
    const align = "align" in style ? style.align : cell.type === "number" ? "right" : "left";
    return `text-align:${align};
        line-height:${height - 1.5}px;`;
  }

  // ---------------------------------------------------------------------------
  // Handlers
  // ---------------------------------------------------------------------------

  processArrowKeys(ev: KeyboardEvent) {
    if (this.getters.getEditionMode() === "selecting") {
      ev.preventDefault();
      return;
    }
    ev.stopPropagation();
    const autoCompleteComp = this.autoCompleteRef.comp as TextValueProvider;
    if (
      ["ArrowUp", "ArrowDown"].includes(ev.key) &&
      this.autoCompleteState.showProvider &&
      autoCompleteComp
    ) {
      ev.preventDefault();
      if (ev.key === "ArrowUp") {
        autoCompleteComp.moveUp();
      } else {
        autoCompleteComp.moveDown();
      }
    }
  }

  processTabKey(ev: KeyboardEvent) {
    ev.preventDefault();
    ev.stopPropagation();
    const autoCompleteComp = this.autoCompleteRef.comp as TextValueProvider;
    if (this.autoCompleteState.showProvider && autoCompleteComp) {
      const autoCompleteValue = autoCompleteComp.getValueToFill();
      if (autoCompleteValue) {
        this.autoComplete(autoCompleteValue);
        return;
      }
    } else {
      // when completing with tab, if there is no value to complete, the active cell will be moved to the right.
      // we can't let the model think that it is for a ref selection.
      // todo: check if this can be removed someday
      this.dispatch("STOP_COMPOSER_SELECTION");
    }

    const deltaX = ev.shiftKey ? -1 : 1;
    this.isDone = true;
    this.dispatch("MOVE_POSITION", { deltaX, deltaY: 0 });
  }

  processEnterKey(ev: KeyboardEvent) {
    ev.preventDefault();
    ev.stopPropagation();
    const autoCompleteComp = this.autoCompleteRef.comp as TextValueProvider;
    if (this.autoCompleteState.showProvider && autoCompleteComp) {
      const autoCompleteValue = autoCompleteComp.getValueToFill();
      if (autoCompleteValue) {
        this.autoComplete(autoCompleteValue);
        return;
      }
    }
    this.dispatch("STOP_EDITION");
    this.dispatch("MOVE_POSITION", {
      deltaX: 0,
      deltaY: ev.shiftKey ? -1 : 1,
    });
    this.isDone = true;
  }

  processEscapeKey() {
    this.dispatch("STOP_EDITION", { cancel: true });
    this.isDone = true;
  }

  onKeydown(ev: KeyboardEvent) {
    let handler = this.keyMapping[ev.key];
    if (handler) {
      return handler.call(this, ev);
    }
    ev.stopPropagation();
  }

  /*
   * Triggered automatically by the content-editable between the keydown and key up
   * */
  onInput(ev: KeyboardEvent) {
    if (this.isDone || !this.shouldProcessInputEvents) {
      return;
    }
    const el = this.composerRef.el! as HTMLInputElement;
    if (el.clientWidth !== el.scrollWidth) {
      el.style.width = `${el.scrollWidth + 20}px`;
    }
    const content = el.childNodes.length ? el.textContent! : "";
    this.dispatch("SET_CURRENT_CONTENT", { content });
  }

  onKeyup(ev: KeyboardEvent) {
    if (
      this.isDone ||
      [
        "Control",
        "Shift",
        "ArrowUp",
        "ArrowDown",
        "ArrowLeft",
        "ArrowRight",
        "Tab",
        "Enter",
      ].includes(ev.key)
    ) {
      // already processed in keydown
      return;
    }
    ev.preventDefault();
    ev.stopPropagation();

    // reset the state of the ref selector and autocomplete for safety.
    // They will be set correctly if needed in `processTokenAtCursor`
    this.autoCompleteState.showProvider = false;
    this.autoCompleteState.search = "";
    this.dispatch("STOP_COMPOSER_SELECTION");
    if (ev.ctrlKey && ev.key === " ") {
      this.autoCompleteState.showProvider = true;
    } else {
      this.processContent();
      this.processTokenAtCursor();
    }
  }

  onClick(ev: MouseEvent) {
    ev.stopPropagation();
    this.processContent();
    this.processTokenAtCursor();
  }
  onCompleted(ev: CustomEvent) {
    this.autoComplete(ev.detail.text);
  }

  // ---------------------------------------------------------------------------
  // Private
  // ---------------------------------------------------------------------------

  processContent() {
    this.shouldProcessInputEvents = false;
    let value = this.getters.getCurrentContent();
    this.tokenAtCursor = undefined;
    if (value.startsWith("=")) {
      this.saveSelection();
      this.contentHelper.removeAll(); // remove the content of the composer, to be added just after
      this.contentHelper.selectRange(0, 0); // move the cursor inside the composer at 0 0.
      this.dispatch("REMOVE_ALL_HIGHLIGHTS"); //cleanup highlights for references

      const refUsed = {};
      let lastUsedColorIndex = 0;

      this.tokens = composerTokenize(value);
      if (this.selectionStart === this.selectionEnd && this.selectionEnd === 0) {
        this.tokenAtCursor = undefined;
      } else {
        this.tokenAtCursor = this.tokens.find(
          (t) => t.start <= this.selectionStart! && t.end >= this.selectionEnd!
        );
      }
      for (let token of this.tokens) {
        switch (token.type) {
          case "OPERATOR":
          case "NUMBER":
          case "FUNCTION":
          case "COMMA":
          case "BOOLEAN":
            this.contentHelper.insertText(token.value, tokenColor[token.type]);
            break;
          case "SYMBOL":
            let value = token.value;
            const [xc, sheet] = value.split("!").reverse();
            if (rangeReference.test(xc)) {
              const refSanitized =
                getComposerSheetName(
                  sheet
                    ? `${sheet}`
                    : `${this.getters.getSheetName(this.getters.getEditionSheet())}`
                ) +
                "!" +
                xc.replace(/\$/g, "");
              if (!refUsed[refSanitized]) {
                refUsed[refSanitized] = colors[lastUsedColorIndex];
                lastUsedColorIndex = ++lastUsedColorIndex % colors.length;
              }
              this.contentHelper.insertText(value, refUsed[refSanitized]);
            } else {
              this.contentHelper.insertText(value);
            }
            break;
          case "LEFT_PAREN":
          case "RIGHT_PAREN":
            // Compute the matching parenthesis
            if (
              this.tokenAtCursor &&
              ["LEFT_PAREN", "RIGHT_PAREN"].includes(this.tokenAtCursor.type) &&
              this.tokenAtCursor.parenIndex &&
              this.tokenAtCursor.parenIndex === token.parenIndex
            ) {
              this.contentHelper.insertText(token.value, MatchingParenColor);
            } else {
              this.contentHelper.insertText(token.value);
            }
            break;
          default:
            this.contentHelper.insertText(token.value);
            break;
        }
      }

      // Put the cursor back where it was
      this.contentHelper.selectRange(this.selectionStart, this.selectionEnd);
      if (Object.keys(refUsed).length) {
        this.dispatch("ADD_HIGHLIGHTS", { ranges: refUsed });
      }
    }
    this.shouldProcessInputEvents = true;
  }

  /**
   * Compute the state of the composer from the tokenAtCursor.
   * If the token is a bool, function or symbol we have to initialize the autocomplete engine.
   * If it's a comma, left_paren or operator we have to initialize the range selection.
   */
  processTokenAtCursor() {
    if (!this.tokenAtCursor) {
      return;
    }
    if (["BOOLEAN", "FUNCTION", "SYMBOL"].includes(this.tokenAtCursor.type)) {
      if (this.tokenAtCursor.value.length > 0) {
        this.autoCompleteState.search = this.tokenAtCursor.value;
        this.autoCompleteState.showProvider = true;
      }
    } else if (["COMMA", "LEFT_PAREN", "OPERATOR", "SPACE"].includes(this.tokenAtCursor.type)) {
      // we need to reset the anchor of the selection to the active cell, so the next Arrow key down
      // is relative the to the cell we edit
      this.dispatch("START_COMPOSER_SELECTION");
      // We set this variable to store the start of the selection, to allow
      // to replace selections (ex: select twice a cell should only be added
      // once)
      this.refSelectionStart = this.selectionStart;
    }
  }

  addText(text: string) {
    this.contentHelper.selectRange(this.selectionStart, this.selectionEnd);
    this.contentHelper.insertText(text);
    this.selectionStart = this.selectionEnd = this.selectionStart + text.length;
    this.contentHelper.selectRange(this.selectionStart, this.selectionEnd);
  }

  addTextFromSelection() {
    const zone = this.getters.getSelectedZones()[0];
    let selection = this.getters.zoneToXC(zone);
    if (this.refSelectionStart) {
      this.selectionStart = this.refSelectionStart;
    }
    if (this.getters.getEditionSheet() !== this.getters.getActiveSheet()) {
      const sheetName = getComposerSheetName(
        this.getters.getSheetName(this.getters.getActiveSheet())!
      );
      selection = `${sheetName}!${selection}`;
    }
    this.addText(selection);
    this.processContent();
  }

  autoComplete(value: string) {
    this.saveSelection();
    if (value) {
      if (this.tokenAtCursor) {
        let start = this.tokenAtCursor.end;
        let end = this.tokenAtCursor.end;

        if (["SYMBOL", "FUNCTION"].includes(this.tokenAtCursor.type)) {
          start = this.tokenAtCursor.start;
        }

        if (this.autoCompleteState.provider && this.tokens.length) {
          value += "(";

          const currentTokenIndex = this.tokens
            .map((token) => token.start)
            .indexOf(this.tokenAtCursor.start);
          if (currentTokenIndex + 1 < this.tokens.length) {
            const nextToken = this.tokens[currentTokenIndex + 1];
            if (nextToken.type === "LEFT_PAREN") {
              end++;
            }
          }
        }
        this.selectionStart = start;
        this.selectionEnd = end;
      }
      this.addText(value);
    }

    this.autoCompleteState.search = "";
    this.autoCompleteState.showProvider = false;
    this.processContent();
    this.processTokenAtCursor();
  }

  /**
   * Save the current selection
   */
  saveSelection() {
    const selection = this.contentHelper.getCurrentSelection();
    this.selectionStart = selection.start;
    this.selectionEnd = selection.end;
  }
}
