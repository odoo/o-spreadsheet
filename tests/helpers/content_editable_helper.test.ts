import { ContentEditableHelper } from "../../src/components/composer/content_editable_helper";
import { iterateChildren } from "../../src/components/helpers/dom_helpers";
import { makeTestFixture } from "../test_helpers/helpers";

/**
 * Mock innerText as it is not implemented in jsDom
 * See https://github.com/jsdom/jsdom/issues/1245
 *
 * This mock is not reflecting a 100% the reality
 * (textContent differs from innerText [1])
 *
 * As we never input the newline character directly in the
 * ContentEditableHelper (we provide an array of paragraphs),
 * The two properties are equivalent and can be 'safely' mocked.
 *
 * [1] https://kellegous.com/j/2013/02/27/innertext-vs-textcontent/
 */
Object.defineProperty(HTMLElement.prototype, "innerText", {
  get: function () {
    return this.textContent;
  },
  set: function (text) {
    this.textContent = text;
  },
});

describe("ContentEditableHelper", () => {
  describe("setText only applies a diff to the current content", () => {
    let div: HTMLDivElement;
    let helper: ContentEditableHelper;
    let fixture: HTMLElement;
    beforeEach(() => {
      fixture = makeTestFixture();
      div = document.createElement("div");
      fixture.appendChild(div);
      div.contentEditable = "true";
      helper = new ContentEditableHelper(div);
    });

    afterEach(() => {
      fixture.remove();
    });

    test("writing text in empty div", () => {
      expect(helper.getText()).toBe("");
      helper.setText([[{ value: "hello" }]]);
      expect(helper.getText()).toBe("hello");
      expect(div.innerText).toBe("hello");
    });

    test("Change color of an HTML content", () => {
      helper.setText([[{ value: "hello", color: "#000000" }]]);
      expect(helper.getText()).toBe("hello");
      expect(div.innerText).toBe("hello");
      const childNodes = [...iterateChildren(div)];
      helper.setText([[{ value: "hello", color: "#ffffff" }]]);
      expect(helper.getText()).toBe("hello");
      expect(div.innerText).toBe("hello");
      const childNodesAfter = [...iterateChildren(div)];
      // span nodes are not preserved
      expect(childNodesAfter).not.toEqual(childNodes);
      expect(childNodesAfter).not.toEqual(childNodes);
    });

    test("Adding some text", () => {
      helper.setText([[{ value: "hello" }]]);

      expect(helper.getText()).toBe("hello");
      expect(div.innerText).toBe("hello");
      const childNodes = [...iterateChildren(div)];
      expect(childNodes.length).toBe(4);
      const [, paragraphNode, spanNode] = childNodes;
      expect(div).toMatchSnapshot();

      helper.setText([[{ value: "hello" }, { value: "test" }]]);
      expect(helper.getText()).toBe("hellotest");
      expect(div.innerText).toBe("hellotest");

      const childNodesAfter = [...iterateChildren(div)];
      expect(childNodesAfter).not.toEqual(childNodes);
      expect(childNodesAfter.length).toBe(6);
      expect(div).toMatchSnapshot();

      const [, paragraphNodeAfter, spanNodeAfter] = childNodesAfter;
      // paragraph node is preserved
      expect(paragraphNodeAfter.nodeName).toBe("P");
      expect(paragraphNodeAfter).toEqual(paragraphNode);
      // the first 'SPAN' node is preserved
      expect(spanNodeAfter.nodeName).toBe("SPAN");
      expect(spanNodeAfter).toEqual(spanNode);
    });

    test("Removing some text", () => {
      helper.setText([[{ value: "hello" }, { value: "test" }]]);

      expect(helper.getText()).toBe("hellotest");
      expect(div.innerText).toBe("hellotest");
      const childNodes = [...iterateChildren(div)];
      expect(childNodes.length).toBe(6);
      const [, paragraphNode, spanNode] = childNodes;
      expect(div).toMatchSnapshot();

      helper.setText([[{ value: "hello" }]]);
      expect(helper.getText()).toBe("hello");
      expect(div.innerText).toBe("hello");

      const childNodesAfter = [...iterateChildren(div)];
      expect(childNodesAfter).not.toEqual(childNodes);
      expect(childNodesAfter.length).toBe(4);
      const [, paragraphNodeAfter, spanNodeAfter] = childNodesAfter;
      expect(div).toMatchSnapshot();

      // paragraph node is preserved
      expect(paragraphNodeAfter.nodeName).toBe("P");
      expect(paragraphNodeAfter).toEqual(paragraphNode);
      // the first 'SPAN' node is preserved
      expect(spanNodeAfter.nodeName).toBe("SPAN");
      expect(spanNodeAfter).toEqual(spanNode);
    });
  });
});
