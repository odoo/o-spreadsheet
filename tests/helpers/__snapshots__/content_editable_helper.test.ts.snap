// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentEditableHelper setText only applies a diff to the current content Adding some text 1`] = `
<div>
  <p>
    <span>
      hello
    </span>
  </p>
</div>
`;

exports[`ContentEditableHelper setText only applies a diff to the current content Adding some text 2`] = `
<div>
  <p>
    <span>
      hello
    </span>
    <span>
      test
    </span>
  </p>
</div>
`;

exports[`ContentEditableHelper setText only applies a diff to the current content Removing some text 1`] = `
<div>
  <p>
    <span>
      hello
    </span>
    <span>
      test
    </span>
  </p>
</div>
`;

exports[`ContentEditableHelper setText only applies a diff to the current content Removing some text 2`] = `
<div>
  <p>
    <span>
      hello
    </span>
  </p>
</div>
`;
